### ISSUE TITLE
netlink: consider memoizing the largest seen buffer size in Conn.Receive

### ISSUE THREAD

**Issue by mdlayher** [mdlayher/netlink#179]
Right now we allocate a page of memory, peek into the socket, and double the size of the buffer if we can't fit the entire message.

For APIs which frequently return a large amount of data, it probably makes sense to memoize the largest used buffer on a per-Conn basis so that we know how much to allocate up front rather than looping with peek/allocate.

**Comment by mdlayher** [comment]
Before committing to doing this with a public API, I may do something along the lines of `NLDEBUG=experiments=conn-memoize-receive` or similar so this can be quickly toggled with no code changes in deployments.

**Comment by helios741** [comment]
I used netlink for internal services and optimized [conn.Receive](https://github.com/mdlayher/netlink/blob/main/conn_linux.go#L120) with sync.Pool, saving GCMark time and 28% CPU savings.
If possible, I can submit a PR to optimize conn.Receive.

Before optimisation:
CPU:
![image](https://user-images.githubusercontent.com/12036324/165299982-2ce16669-5bc5-4a9c-a923-207aa7b6b4d9.png)

mem:
![image](https://user-images.githubusercontent.com/12036324/165300047-e4732e9b-0508-4e94-bc1d-0beb3224b002.png)

After optimization:
CPU:
![image](https://user-images.githubusercontent.com/12036324/165300164-f1814a2f-f2e3-4bbb-96a6-08055ae01e55.png)

Mem:
![image](https://user-images.githubusercontent.com/12036324/165300269-1aa1dfc5-3897-48e3-ab70-0369393ddf48.png)

**Comment by mdlayher** [comment]
I don't believe a sync.Pool optimization would be safe because the library hands out allocated buffers for caller use, and putting those back in the pool could mean they are clobbered.

I think a more appropriate approach to reuse memory would be to add an API which reads into a caller-allocated buffer and leaves managing that up to the caller.
